src/machi_file_proxy.erl

```diff
@@ -529,20 +578,27 @@ check_or_make_tagged_csum(OtherTag, _ClientCsum, _Data) ->
 % </li>
 %
 % On success, `{ok, Bytes, Checksum}' is returned.
-handle_read(FHd, Filename, undefined, Offset, Size, U) ->
-    handle_read(FHd, Filename, machi_util:make_tagged_csum(none), Offset, Size, U);
-
-handle_read(FHd, Filename, TaggedCsum, Offset, Size, U) ->
-    case is_byte_range_unwritten(Offset, Size, U) of
-        true ->
+handle_read(FHd, Filename, CsumTable, Offset, Size, U) ->
+    case {is_byte_range_unwritten(Offset, Size, U),
+          machi_csum_table:find(CsumTable, Offset, Size)} of
+        {true, _} ->
+            %% More close look might be needed to care result of csum
+            %% table, but for now this just works for known chunks
+            %% (TODO^^;).
             {error, not_written};
-        false ->
+        {false, {error, trimmed} = E} ->
+            E;
+        {false, {ok, TaggedCsum}} ->
+            %% Something is wrong :S
+            do_read(FHd, Filename, TaggedCsum, Offset, Size);
+        {false, {error, _}} ->
```

Comment:
Hrm.  There is a similar problem for read when there is overlap of written and not-written bytes.

I think perhaps that I should change the return type of read_chunk() ... the result will be more complicated, but ... I think that the complication is going to be necessary.

Right now:

    message Mpb_ReadChunkResp {
        required Mpb_GeneralStatusCode status = 1;
        optional bytes chunk = 2;
        optional Mpb_ChunkCSum csum = 3;
    }

Perhaps it should be something like:

    message Mpb_ReadChunkResp {
        required Mpb_GeneralStatusCode status = 1;
        repeated optional bytes chunks = 2;     /* not sure if this is valid syntax */
        repeated optional uint64 offsets = 3     /* not sure if this is valid syntax */
        repeated optional Mpb_ChunkCSum csums = 5;     /* not sure if this is valid syntax */
    }

Hrm, that doesn't tell the client the difference between not-written and trimmed.  Thoughts?